conbencher/python/conbencher/adapters/gbench.py

```diff
@@ -0,0 +1,226 @@
+    """
+
+    def __init__(
+        self,
+        name: str,
+        real_time: float,
+        cpu_time: float,
+        time_unit: str,
+        run_type: str,
+        size=None,
+        bytes_per_second: float = None,
+        items_per_second: float = None,
+        **counters: dict,
+    ):
+        self._name = name
+        self.real_time = real_time
+        self.cpu_time = cpu_time
+        self.time_unit = time_unit
+        self.run_type = run_type
+        self.size = size
+        self.bytes_per_second = bytes_per_second
+        self.items_per_second = items_per_second
+        self.counters = counters
+
+    @property
+    def is_aggregate(self):
+        """Indicate if the observation is a run or an aggregate."""
+        return self.run_type == "aggregate"
+
+    @property
+    def is_realtime(self):
+        """Indicate if the preferred value is realtime instead of cputime."""
+        return self.name.find("/real_time") != -1
+
+    @property
+    def name(self):
+        name = self._name
+        return name.rsplit("_", maxsplit=1)[0] if self.is_aggregate else name
+
+    @property
+    def time(self):
+        return self.real_time if self.is_realtime else self.cpu_time
+
+    @property
+    def value(self):
+        """Return the benchmark value."""
+        return self.bytes_per_second or self.items_per_second or self.time
+
+    @property
+    def unit(self):
+        if self.bytes_per_second:
+            return "bytes_per_second"
+        elif self.items_per_second:
+            return "items_per_second"
+        else:
+            return self.time_unit
+
+
+@dataclass
+class GoogleBenchmark:
+    """A set of GoogleBenchmarkObservations."""
+
+    name: str
+    unit: str
+    time_unit: str
+    less_is_better: bool
+    values: list[float]
+    times: list[float]
+    counters: list = None
+
+    @classmethod
+    def from_runs(cls, name: str, runs: list[GoogleBenchmarkObservation]):
+        """
+        Create a GoogleBenchmarkGroup instance from a list of observations
+
+        Parameters
+        ----------
+        name: str
+              Name of the benchmark
+        runs: list(GoogleBenchmarkObservation)
+              Repetitions of GoogleBenchmarkObservation run.
+        """
+        return cls(
+            name=name,
+            unit=runs[0].unit,
+            time_unit=runs[0].time_unit,
+            less_is_better=not runs[0].unit.endswith("per_second"),
+            values=[b.value for b in runs],
+            times=[b.real_time for b in runs],
+        )
+
+
+class GoogleBenchmarkAdapter(_BenchmarkAdapter):
+    """A class for running Google Benchmarks and sending the results to conbench"""
+
+    def __init__(self, command: list[str]) -> None:
+        """
+        Parameters
+        ----------
+        command : list[str]
+            A list of strings defining a shell command to run the benchmarks
+        """
+        self.result_file = NamedTemporaryFile().name
```

Comment:
I might be completely wrong here, but I was under the impression that if you did something like
```
file = NamedTemporaryFile()
filename = file.name
```
then the file wouldn't be cleaned up until the `file` variable is out of the scope of the interpreter and garbage-collected. So when you do `NamedTemporaryFile().name`, since there's no variable storing the `NamedTemporaryFile` object, it could be garbage-collected sooner? To be safe, you should put that object in a separate variable.

Reply:
Oh yeah I just validated this is the case.
```
>>> import tempfile
>>> x = tempfile.NamedTemporaryFile().name
>>> x
'/var/folders/fz/ydy9vm4d6qbgxw11ys6j_kfh0000gn/T/tmpq2myi2ed'
>>> with open(x) as f:
...  print(f.read())
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/fz/ydy9vm4d6qbgxw11ys6j_kfh0000gn/T/tmpq2myi2ed'
>>>
>>> file = tempfile.NamedTemporaryFile()
>>> x = file.name
>>> with open(x) as f:
...  print(f.read())
...

>>>
```

Reply:
It's mostly just to generate a path here—the actual file needs to be written by the benchmark, not this class. Actually, I think maybe we need to make users specify it; the `--output` bit below this was left over from archery

Reply:
Ok, fixed this, and where we do need tempfiles, switched to `tempfile.mktemp()`, because the context manager was just distracting here